File: easy/python/implement_queue_using_stacks.py
from ast import List


class MyQueue:
    def __init__(self):
        self.stack1: List[int] = []
        self.stack2: List[int] = []

    def push(self, x: int) -> None:
        self.stack1.append(x)

    def pop(self) -> int:
        self.peek()
        return self.stack2.pop()

    def peek(self) -> int:
        if not self.stack2:
            while self.stack1:
                self.stack2.append(self.stack1.pop())
        return self.stack2[len(self.stack2) - 1]

    def empty(self) -> bool:
        return not len(self.stack1) and not len(self.stack2)


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: easy/ruby/implement_queue_using_stacks.rb
class MyQueue
  def initialize
    @stack1 = []
    @stack2 = []
  end

  #   :type x: Integer
  #   :rtype: Void
  def push(x)
    @stack1.push(x)
  end

  #   :rtype: Integer
  def pop
    move_elements if @stack2.empty?
    @stack2.pop
  end

  #   :rtype: Integer
  def peek
    move_elements if @stack2.empty?
    @stack2.last
  end

  #   :rtype: Boolean
  def empty
    @stack1.empty? && @stack2.empty?
  end

  private

  def move_elements
    @stack2.push(@stack1.pop) while @stack1.any?
  end
end

# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue.new()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()
